source/kernel/src/registry/mod.rs

```diff
@@ -0,0 +1,440 @@
+                    },
+                    &mut wgr,
+                )
+                .map_err(drop)?;
+                let len = used.len();
+                wgr.commit(len);
+                Ok(())
+            }
+        }
+    }
+}
+
+// UserspaceHandle
+
+impl UserspaceHandle {
+    pub fn process_msg(
+        &self,
+        user_msg: UserRequest<'_>,
+        user_ring: &bbq::MpscProducer,
+    ) -> Result<(), ()> {
+        unsafe { (self.req_deser)(user_msg, &self.req_producer_leaked, user_ring) }
+    }
+}
+
+// KernelHandle
+
+impl<RD: RegisteredDriver> KernelHandle<RD> {
+    pub async fn send(&self, msg: RD::Request, reply: ReplyTo<RD::Response>) -> Result<(), ()> {
+        self.prod
+            .enqueue_async(Message {
+                msg: Envelope { body: msg },
+                reply,
+            })
+            .await
+            .map_err(drop)
+    }
+}
+
+// -- other --
+
+/// A monomorphizable function that allows us to store the serialization type within
+/// the function itself, allowing for a type-erased function pointer to be stored
+/// inside of the registry.
+unsafe fn map_deser<RD>(
+    umsg: UserRequest<'_>,
+    req_tx: &LeakedKProducer,
+    user_resp: &bbq::MpscProducer,
+) -> Result<(), ()>
+where
+    RD: RegisteredDriver,
+    RD::Request: Serialize + DeserializeOwned,
+    RD::Response: Serialize + DeserializeOwned,
+{
+    // Un-type-erase the producer channel
+    //
+    // TODO: We don't really need to clone the producer, we just need a reference valid
+    // for the lifetime of `req_tx`. Consider adding a method for this before merging
+    // https://github.com/tosc-rs/mnemos/pull/25.
+    //
+    // This PROBABLY would require a "with"/closure method to make sure the producer ref
+    // doesn't outlive the LeakedKProducer reference.
+    let req_prod = req_tx.clone_typed::<Message<RD>>();
+
+    // Deserialize the request, if it doesn't have the right contents, deserialization will fail.
+    let u_payload: RD::Request = postcard::from_bytes(umsg.req_bytes).map_err(drop)?;
+
+    // Create the message type to be sent on the channel
+    let msg: Message<RD> = Message {
+        msg: Envelope { body: u_payload },
+        reply: ReplyTo::Userspace {
+            nonce: umsg.nonce,
+            outgoing: user_resp.clone(),
+        },
+    };
+
+    // Send the message, and report any failures
+    req_prod.enqueue_sync(msg).map_err(drop)
+}
+
+/// TODO: I don't really know what to do with this. This is essentially
+/// declaring the client interface of a "Simple Serial" driver, without
+/// providing a definition of the actual driver service. In the simulator,
+/// this interface is used for the TcpSerial driver. This sort of "forward
+/// declaration" is needed when a driver in the kernel (like the SerialMux)
+/// depends on some external definition.
+///
+/// For non-kernel-depended services, it should be enough to depend on the
+/// actual driver you are consuming
+pub mod simple_serial {
```

Comment:
nit, tioli: should this maybe go in a `kernel::drivers` module or something, instead of in `registry`?